ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java

```diff
@@ -1056,6 +1061,72 @@ private void registerCommandTopic() {
 
     final String commandTopic = commandStore.getCommandTopicName();
 
+    // migration code
+    if (checkMigrationConditions(commandTopic)) {
+      log.warn("Migrating command topic from the service context Kafka to the command "
+          + "producer/consumer Kafka.");
+      KsqlInternalTopicUtils.ensureTopic(
+          commandTopic,
+          ksqlConfigNoPort,
+          internalTopicClient
+      );
+
+      try {
+        CommandTopicMigrationUtil.commandTopicMigration(commandTopic, restConfig, ksqlConfigNoPort);
+      } catch (final Exception e) {
+        throw new KsqlException("Error migrating command topic", e);
+      }
+    } else if (restConfig.getString(KsqlRestConfig.KSQL_COMMAND_TOPIC_MIGRATION_CONFIG)
+        .equals(KsqlRestConfig.KSQL_COMMAND_TOPIC_MIGRATION_MIGRATING)) {
+      final TopicPartition topicPartition = new TopicPartition(commandTopic, 0);
+
+      final org.apache.kafka.clients.consumer.Consumer<byte[], byte[]> oldBrokerConsumer =
+          new KafkaConsumer<>(
+              restConfig.originals(),
+              new ByteArrayDeserializer(),
+              new ByteArrayDeserializer()
+          );
+      oldBrokerConsumer.assign(Collections.singleton(topicPartition));
+
+      final Map<String, Object> newConsumerConfig = restConfig.getCommandConsumerProperties();
+      final org.apache.kafka.clients.consumer.Consumer<byte[], byte[]> newBrokerConsumer =
+          new KafkaConsumer<>(
+              newConsumerConfig,
+              new ByteArrayDeserializer(),
+              new ByteArrayDeserializer()
+          );
+      newBrokerConsumer.assign(Collections.singleton(topicPartition));
+
+      RetryUtil.retryWithBackoff(
```

Comment:
the old and new may not actually have the same number of commands, even after the migration in the event that some command slips in behind the "bad" command. Can't we just check whether or not any record is in the new command topic? The writes to it are transactional, so we either see the whole migration or no migration.

Reply:
Sure i'll do that, that'll simplify this part